spinalcordtoolbox/image.py

```diff
@@ -411,9 +413,23 @@ def loadFromPath(self, path, mmap, verbose):
 
         self.absolutepath = os.path.abspath(path)
         im_file = nib.load(self.absolutepath, mmap=mmap)
-        self.affine = im_file.affine.copy()
-        self.data = im_file.get_data()
         self.hdr = im_file.header.copy()
+        self.affine = im_file.affine.copy()
+
+        # Check to see if there's a mismatch between the array's datatype and the header datatype
+        dtype_apparent = np.result_type(im_file.dataobj)
+        dtype_header = self.hdr.get_data_dtype()
+        if dtype_apparent != dtype_header:
+            logger.warning(f"DTYPEWARN: Image header specifies datatype {dtype_header}, but array appears to be of "
+                           f"type {dtype_apparent}. Array will be loaded as type {dtype_header}.")
+        # Convert the data array to the header's datatype
+        self.data = im_file.get_fdata().astype(dtype_header)
```

Comment:
In the above changes, I attempt to cast the data array to the type specified by the header.

When I first made these changes, I had hoped that the header's datatype property would be a useful source of truth, and that it would help to dictate what type the array "should" have. 

In reality, however:

1. In many cases, the header metadata is not to be trusted. For example:
    - In SCT's codebase, we often copy headers from one image to another (`im = Image(data, old_header)`) to preserve metadata.
    - We will also often modify data arrays directly (`im.data = updated_array`) without updating the header. 
    - This means that the header datatype property will often be outdated and incorrect.
2. In addition, casting the array datatype using `.astype(dtype_header)` is a very destructive operation, because it can often impose an incorrect datatype on an otherwise OK array.

So, I think instead, it's much safer to treat the _array_ as the source of truth, and update the _header_ on save/load to reflect the datatype of the array.

Reply:
> So, I think instead, it's much safer to treat the array as the source of truth, and update the header on save/load to reflect the datatype of the array.

Upon making these changes, we now only have 2 failing tests:

```
_______________________________________________________________ test_sct_dmri_moco_check_params ________________________________________________________________
Traceback (most recent call last):
  File "/home/joshua/repos/spinalcordtoolbox/testing/cli/test_cli_sct_dmri_moco.py", line 21, in test_sct_dmri_moco_check_params
    assert allclose(lresults, lgroundtruth)
AssertionError: assert False
 +  where False = allclose(array([ 4.75290417e-04, -1.19489128e-05, -1.19489128e-05, -1.19489128e-05,\n       -1.29636934e-01, -1.29636934e-01, -1.29636934e-01]), 
                                 [0.00047529041677414337, -1.1970542445283172e-05, -1.1970542445283172e-05, -1.1970542445283172e-05, -0.1296642741802682, -0.1296642741802682, ...])
__________________________________________________________ test_sct_dmri_moco_with_mask_check_params ___________________________________________________________
Traceback (most recent call last):
  File "/home/joshua/repos/spinalcordtoolbox/testing/cli/test_cli_sct_dmri_moco.py", line 44, in test_sct_dmri_moco_with_mask_check_params
    assert allclose(lresults, lgroundtruth)
AssertionError: assert False
 +  where False = allclose(array([ 0.00804089,  0.00395558,  0.00395558,  0.00395558, -0.01534442,\n       -0.01534442, -0.01534442]), [0.008032332623754357, 0.0037734940916436697, 0.0037734940916436697, 0.0037734940916436697, -0.01502861167728611, -0.01502861167728611, ...])
```

The differences are tiny in magnitude; I'm fairly certain these failures are due to the fact that we're hardcoding these parameters.

I've had reservations about these tests before:
> If this check fails, though, it doesn't give us much qualitative feedback:
> 
> * The test fails in an identical way for changes of any magnitude (e.g. 1e-4 and 1e2), so it's hard to gauge the severity of a failing test.
> * The test displays the moco params, but doesn't provide any message describing the values, so it's not immediately obvious what it means for these values to change.
>  * Checking against these values only tests if we're preserving the current state of the moco algorithm. But, I think it would be valuable to also try to capture the goal of motion correction, too. (e.g. Say we started with misaligned data, and had ground truth for aligned data. Then we could check if the moco algorithm performs the realignment to an acceptable degree.)
>
> _Source: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3075#issuecomment-871366336_

So, I'm not concerned that this test is failing.

Reply:
tl;dr: I think the failing test is actually due to the current "ground truth" values being wrong (due to corruption from scaling issues). And, I think the new values are actually the correct ones!

<details> 
<summary> Investigation </summary>

I set breakpoints within both of the new datatype mismatch checks. 

Immediately, I found `float/int` mismatches on these lines

  * https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/d89f8a3de5253cdca5c5041a467422aa8214fa24/spinalcordtoolbox/moco.py#L244
  
    https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/d89f8a3de5253cdca5c5041a467422aa8214fa24/spinalcordtoolbox/moco.py#L283
  
    (The `.mean` call causes the `int16`-typed `im` array to change to `float64`)
  
  * https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/d89f8a3de5253cdca5c5041a467422aa8214fa24/spinalcordtoolbox/moco.py#L612-L614

    (The mathematical operation causes the `int16`-typed `data_targetz` array to change to `float32`.)
  
  * https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/d89f8a3de5253cdca5c5041a467422aa8214fa24/spinalcordtoolbox/moco.py#L377-L378
  
    (Originally, `im_moco` has a `float64` header. However, `im_data` has an `int16` header which gets copied over.)

In all four cases, having a `float` array and an `int` header is what what triggers: 

* #3232. 
 
So, it is likely that the values were being slightly corrupted! 

However, after this PR, the images are saved with `float64` in both the array _and_ header, avoiding any scaling issues.

</details>

Reply:
I've rewritten this PR to modify the header, rather than the array.

Reply:
I agree that the array dtype is much more reliable as a source of truth, so updating the header's dtype makes the most sense.